Claim listing

nmrcarroll opened the pull request:
Created the tabbed view we wanted.

Not entirely sure how the elastic search is going to work, right now I've implemented this as pulling claims from two separate files one "claims" the ones made by the user and two "approval" the ones made by anyone other then the user.

nmrcarroll committed:
Partially working tabbed view. Still needs work.

nmrcarroll committed:
Tabs working

nmrcarroll committed:
Re-enable first run check

nmrcarroll committed:
Cleeanup code

nmrcarroll committed:
Reenable first run check

nmrcarroll committed:
Merge branch 'master' into claim_listing

Conflicts:
	Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java

nmrcarroll committed:
Fixed merge conflicts I think.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+    protected static final String EXPENSE_APPROVAL_FILENAME = "approval";
+    private static final String PREFERENCE = "PREFERENCE";
+	private static final int ADD_EXPENSE_CLAIM_REQUEST = 1;
+	private static final int EDIT_EXPENSE_CLAIM_REQUEST = 2;
```
File needs to be reindented

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+    
+    protected ListModel<ExpenseClaim> listModel;
+    
+    ListView myButton;
```
Seems odd to name a list view `myButton`

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+    	super.onActivityCreated(savedInstanceState);
+    	
+    	
+    	final ActionMode.Callback longClickCallback = new ActionMode.Callback() {
```
This should use `LongClickDeleteListener` instead, it eliminates a lot of this boilerplate code

nmrcarroll replied:
Not sure what you mean here, can't find much info on "LongClickDeleteListener".
Did you want to skip the checkmark trashcan selection and go straight to the alert dialogue?

getListView().setOnItemLongClickListener(new OnItemLongClickListener() {

```
        @Override
        public boolean onItemLongClick(AdapterView<?> arg0, View arg1,
                int arg2, long arg3) {
            startDeleteAlertDialog();
            return true;
        }
    });
```

mrthegreensky replied:
I think this is what Indragie meant 

```
        getListView().setOnItemLongClickListener(
            new LongClickDeleteListener(this, new LongClickDeleteListener.OnDeleteListener() {
                @Override
                public void onDelete(int position) {
                    if (getListView().getItemAtPosition(position) == null) {
                        return;
                    }
```

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+    @SuppressWarnings("static-access")
+	@Override
+	public void onActivityResult(int requestCode, int resultCode, Intent data) {
+		if (resultCode != getActivity().RESULT_OK) return;
```
Use `Activity.RESULT_OK` and you can remove the `@SuppressWarnings`

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+		
+	}
+	
+	public void load(){
```
Should this be removed?

nmrcarroll replied:
Not sure which you're trying to highlight here.

indragiek replied:
The `load()` method, since it has nothing in it

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+	private void onSortExpenseResult(Intent data) {
+		Comparator<ExpenseClaim> comparator = (Comparator<ExpenseClaim>)data.getSerializableExtra(ExpenseClaimSortActivity.EXPENSE_CLAIM_SORT);
+		listModel.setComparator(comparator);
+		refresh();
```
`refresh()` shouldn't be called from these `on*` methods. The list model will notify observers when something changes through the `update()` method (which I see is implemented below, but has an empty implementation). `refresh()` should be called inside `update()`.

nmrcarroll replied:
I've tried this implementation, but it seems that update is not called on changes.

Scratch that, by implementing the other changes I've resolved that issue.

indragiek replied:
It's because you called addObserver() on the activity instead of the fragment (I mentioned this in another comment). Once you fix that it will work.

## 

Indragie Karunaratne

On Thu, Apr 2, 2015 at 8:22 PM, Nic Carroll notifications@github.com
wrote:

> > -            break;
> > -        case SORT_EXPENSE_CLAIM_REQUEST:
> > -            onSortExpenseResult(data);
> > -            break;
> > -        }
> > -    }
> > -   
> > -    /**
> > -     \* Changes the sorting mode based on a comparator chosen by {@link ExpenseClaimSortActivity}
> > -     \* @param data The intent to get the comparator from.
> > -     */
> > -    @SuppressWarnings("unchecked")
> > -    private void onSortExpenseResult(Intent data) {
> > -        Comparator<ExpenseClaim> comparator = (Comparator<ExpenseClaim>)data.getSerializableExtra(ExpenseClaimSortActivity.EXPENSE_CLAIM_SORT);
> > -        listModel.setComparator(comparator);
> > -        refresh();
> >   I've tried this implementation, but it seems that update is not called on changes.
> >   ---
> >   Reply to this email directly or view it on GitHub:
> >   https://github.com/CMPUT301W15T09/Team9Project/pull/33/files#r27712573

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	@Override
+	public void update(TypedObservable<CollectionMutation<ExpenseClaim>> o,
```
See comment about `refresh()` above

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ClaimTabFragment.java:
```diff
@@ -0,0 +1,18 @@
+	@Override
+	public void load(){
+		listModel = new ListModel<ExpenseClaim>(EXPENSE_CLAIM_FILENAME, getActivity());
+		listModel.addObserver((TypedObserver<CollectionMutation<ExpenseClaim>>) getActivity());
```
The observer should be the fragment (i.e. `this`), not the activity, since you moved all of the code from the activity into the fragment, including the observer callbacks. You can also remove the `@SuppressWarnings` after changing that.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ApprovalTabFragment.java:
```diff
@@ -0,0 +1,16 @@
+	@Override
+	public void load(){
+		listModel = new ListModel<ExpenseClaim>(EXPENSE_APPROVAL_FILENAME, getActivity());
+		listModel.addObserver((TypedObserver<CollectionMutation<ExpenseClaim>>) getActivity());
```
Repeat from comment below:

The observer should be the fragment (i.e. `this`), not the activity, since you moved all of the code from the activity into the fragment, including the observer callbacks.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ApprovalTabFragment.java:
```diff
@@ -0,0 +1,16 @@
+package com.indragie.cmput301as1;
```
Needs license

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+    	setHasOptionsMenu(true);
+        
+    	
+    	load();
```
This method is empty, so this call can be removed.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+package com.indragie.cmput301as1;
```
Needs license

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+
+ 
+
+public class BaseTabFragment extends ListFragment implements TypedObserver<CollectionMutation<ExpenseClaim>> {
```
Name could be more descriptive, `ExpenseClaimTabFragment` or similar?

indragiek replied:
Class needs documentation

indragiek commented on Team9Project/src/com/indragie/cmput301as1/BaseTabFragment.java:
```diff
@@ -0,0 +1,292 @@
+
+public class BaseTabFragment extends ListFragment implements TypedObserver<CollectionMutation<ExpenseClaim>> {
+ 
+    protected static final String EXPENSE_CLAIM_FILENAME = "claims";
```
Constants need to be documented

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ClaimTabFragment.java:
```diff
@@ -0,0 +1,18 @@
+package com.indragie.cmput301as1;
```
Needs license

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ClaimTabListener.java:
```diff
@@ -0,0 +1,41 @@
+package com.indragie.cmput301as1;
```
Needs license

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ClaimTabListener.java:
```diff
@@ -0,0 +1,41 @@
+import android.app.Fragment;
+import android.app.FragmentTransaction;
+
+public class ClaimTabListener<T extends Fragment> implements TabListener {
```
Class needs docs

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ClaimTabListener.java:
```diff
@@ -0,0 +1,41 @@
+import android.app.FragmentTransaction;
+
+public class ClaimTabListener<T extends Fragment> implements TabListener {
+	private Fragment mFragment;
```
We don't use the `m` prefix in the rest of the codebase so it should be consistent with that. If the code style is like this because it was adapted from some other source, that should be cited.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java:
```diff
@@ -77,47 +60,25 @@ protected void onCreate(Bundle savedInstanceState) {
+
+		/* Creating ANDROID Tab */
+		Tab tab = actionBar.newTab()
+			.setText("Claims")
```
The Claims text should be specified inside strings.xml

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java:
```diff
@@ -18,50 +18,33 @@
  */
-public class ExpenseClaimListActivity extends ListActivity implements TypedObserver<CollectionMutation<ExpenseClaim>> {
+
+public class ExpenseClaimListActivity extends FragmentActivity implements TypedObserver<CollectionMutation<ExpenseClaim>> {
```
This file also needs a reindent

indragiek replied:
This no longer needs to implement `TypedObserver` if all of that code has been implemented inside the fragment

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java:
```diff
@@ -77,47 +60,25 @@ protected void onCreate(Bundle savedInstanceState) {
-			public void onDestroyActionMode(ActionMode mode) {}
+		/* Creating APPLE Tab */
+		tab = actionBar.newTab()
+			.setText("Approval")
```
Same as above (strings inside strings.xml)

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java:
```diff
@@ -271,57 +132,15 @@ public void onClick(DialogInterface dialog, int whichButton) {
 	@Override
-	public void update(TypedObservable<CollectionMutation<ExpenseClaim>> observable, CollectionMutation<ExpenseClaim> mutation) {
-		setListAdapter(new ExpenseClaimArrayAdapter(this, listModel.getItems()));
+	public void update(TypedObservable<CollectionMutation<ExpenseClaim>> o,
```
This can be removed once you remove the `TypedObserver` conformance

indragiek commented:
Left some initial comments, there also needs to be unit tests for this (for the tab UI). Like Karan said, they can be pretty basic.

indragiek commented:
@OriginalFool Any progress on this? Would be best to merge this sooner rather than later because a lot of this stuff is changing on master right now.

nmrcarroll commented:
I've made all the requested changes I think. The only thing left is the test cases, having a hard time figuring out UI tests.

indragiek commented:
@OriginalFool I think you may have forgotten to push up your changes because there are no new commits.

nmrcarroll commented:
Yes I was waiting until I had the testing done before I pushed. I can push when I get home if you'd like.

indragiek commented:
Oh I see. I wonder if there's a way to simulate presses on the tabs and check that they actually switch? I don't know, it seems like we'd just be testing the Android UI controls themselves at that point.

nmrcarroll committed:
Added in test cases and cleaned up code.

nmrcarroll committed:
Merge branch 'master' into claim_listing

Conflicts:
	Team9Project/gen/com/indragie/cmput301as1/R.java
	Team9Project/res/values/strings.xml
	Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java

nmrcarroll committed:
Merge branch 'master' into claim_listing

Conflicts:
	Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java

nmrcarroll committed:
Cleaned up as requested

nmrcarroll committed:
Added on create to ExpenseClaimTabFragment to get the activity.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,316 @@
+			CollectionMutation<ExpenseClaim> arg) {
+		setListAdapter(new ExpenseClaimArrayAdapter(activity,
+				listModel.getItems()));
+		// TODO Auto-generated method stub
```
Remove this comment

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,316 @@
+		setHasOptionsMenu(true);
+
+		getListView().setOnItemLongClickListener(
+				new LongClickDeleteListener(activity, new LongClickDeleteListener.OnDeleteListener() {
```
Still needs indentation fix

indragiek commented:
Couple minor comments, but once you merge master in again this should be good to go :+1:

nmrcarroll committed:
Merge branch 'master' into claim_listing

Conflicts:
	Team9Project/gen/com/indragie/cmput301as1/R.java
	Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java

nmrcarroll committed:
Attempt to fix saving

nmrcarroll commented:
With the merge of master it seems that claims are no longer persistent. Not sure how to go about fixing this.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java:
```diff
@@ -54,11 +37,14 @@
 	//================================================================================
 
 	/**
+<<<<<<< HEAD
```
These merge conflict markers should be removed.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+		super.onActivityCreated(savedInstanceState);
+
+
+		getListView().setOnItemLongClickListener(
```
Indentation is broken

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+	/**
+	 * Loads the expense claim data to display in the {@link ListView}
+	 */
+	protected void loadData(int i) {
```
What is the integer parameter used for?

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+	/**
+	 * Manages the user and associated preferences.
+	 */
+	protected UserManager userManager;
```
This should be part of the activity and not the fragment, since the data is the same for both.

nmrcarroll replied:
The usermanager?

indragiek replied:
Yep

nmrcarroll replied:
So implement it in the activity and then pass it on to the fragment? I do need to use the usermanager within the fragment.

indragiek replied:
Yeah, that would work.

nmrcarroll replied:
I'm just wondering why exactly? UserManager isn't used at all within expenseclaimlistactivity, but is used a fair amount within the fragments.

indragiek replied:
It's because if you use it in both the fragments you're creating two instances of `UserManager` and therefore 2 active users. There can only be a single active user.

nmrcarroll replied:
Okay in the most recent build instead of making a new UserManager in each fragment I've set the userManager to be Activity.getUserManager();

Will that be sufficient, or should I remove userManager entirely and make a bunch of calls to Activity.getUserManager()?

indragiek replied:
That's fine for now, we can always refactor things a bit more later. Let me know when you've made all the changes you want to so we can look over everything again.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+	/**
+	 * Index of a item that is long pressed.
+	 */
+	private int longPressedItemIndex;
```
Unused

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+
+
+		getListView().setOnItemLongClickListener(
+				new LongClickDeleteListener(activity, new LongClickDeleteListener.OnDeleteListener() {
```
Looks like this part wasn't merged correctly? The implementation in master is different.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+	 * Prompts the user for confirmation in response to deleting an expense
+	 * claim.
+	 */
+	public void startDeleteAlertDialog() {
```
Looks like this wasn't merged correctly either, the implementation in master takes an integer parameter

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+		Session.setSharedSession(session);
+
+		// Show the initial list of expense claims (persisted on disk)
+		listModel = session.getOwnedClaims();
```
Pretty much everything from here and below needs to be in the separate tab fragments. `getOwnedClaims()` returns the list model for the user's claims. There's another method, something like `getReviewalClaims()` that returns the list model for claims to be approved by the user.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+
+		//Load the new list from the server
+		final Context context = activity;
+		session.loadOwnedClaims(new ElasticSearchAPIClient.APICallback<List<ExpenseClaim>>() {
```
Same here, there's another method on `Session` for loading the claims for approval by the user.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ExpenseClaimTabFragment.java:
```diff
@@ -0,0 +1,482 @@
+		final Context context = activity;
+		session.loadOwnedClaims(new ElasticSearchAPIClient.APICallback<List<ExpenseClaim>>() {
+			@Override
+			public void onSuccess(Response response, List<ExpenseClaim> model) {}
```
This shouldn't be empty. This is likely why you're seeing that it's not persisting. Check the implementation in master.

nmrcarroll replied:
I assume you mean "public void onSuccess(Response response, List<ExpenseClaim> model) {}"
That is actually how it is in master.

indragiek replied:
Whoops, you're right. I don't know when that changed, it must've been one of the merges. I'll fix it right now.

indragiek replied:
I think this should be OK now. You might've noticed some weird behavior while I was fixing it. Merge in master again and check.

nmrcarroll committed:
Updating as requested.

nmrcarroll committed:
Also fixed other errors. Merge branch 'master' into claim_listing

Conflicts:
	Team9Project/src/com/indragie/cmput301as1/ExpenseClaimListActivity.java

nmrcarroll committed:
Removed comment.

nmrcarroll commented:
I think this should be everything now.

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ApprovalTabFragment.java:
```diff
@@ -0,0 +1,126 @@
+	private void loadData() {
+		// Create the application-wide session
+		Session session = new Session(activity, userManager.getActiveUser());
+		Session.setSharedSession(session);
```
This should only be set up once, in the activity. (both lines)

indragiek commented on Team9Project/src/com/indragie/cmput301as1/ClaimTabFragment.java:
```diff
@@ -0,0 +1,127 @@
+	private void loadData() {
+		// Create the application-wide session
+		Session session = new Session(activity, userManager.getActiveUser());
+		Session.setSharedSession(session);
```
Same here

nmrcarroll committed:
Move session back to activity.

indragiek commented:
Looks OK now, but it would be good if we merged this last because it changes a bunch of other stuff.

indragiek commented:
The app crashes as soon as I try to run it from this branch.

indragiek commented:
Here's the exception trace:

```
04-06 20:41:58.910: E/AndroidRuntime(12780): FATAL EXCEPTION: main
04-06 20:41:58.910: E/AndroidRuntime(12780): java.lang.RuntimeException: Unable to start activity ComponentInfo{com.indragie.cmput301as1/com.indragie.cmput301as1.ExpenseClaimListActivity}: java.lang.NullPointerException
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2180)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2230)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.ActivityThread.access$600(ActivityThread.java:141)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1234)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.os.Handler.dispatchMessage(Handler.java:99)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.os.Looper.loop(Looper.java:137)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.ActivityThread.main(ActivityThread.java:5041)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at java.lang.reflect.Method.invokeNative(Native Method)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at java.lang.reflect.Method.invoke(Method.java:511)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:793)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:560)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at dalvik.system.NativeStart.main(Native Method)
04-06 20:41:58.910: E/AndroidRuntime(12780): Caused by: java.lang.NullPointerException
04-06 20:41:58.910: E/AndroidRuntime(12780):    at com.indragie.cmput301as1.ClaimTabFragment.loadData(ClaimTabFragment.java:67)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at com.indragie.cmput301as1.ClaimTabFragment.onCreate(ClaimTabFragment.java:52)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.Fragment.performCreate(Fragment.java:1673)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:854)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1057)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.BackStackRecord.run(BackStackRecord.java:682)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1435)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.Activity.performStart(Activity.java:5113)
04-06 20:41:58.910: E/AndroidRuntime(12780):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2153)
04-06 20:41:58.910: E/AndroidRuntime(12780):    ... 11 more
```

nmrcarroll commented:
Yeah, I'm having an issue with that. It was because user wasn't defined yet to create the session. The issue I'm having is the fragments start before I can load the session even doing it through the activity. It will work on second launch, just not first.

nmrcarroll committed:
Crashes first use. Will any use after.

indragiek commented:
Alright, I'm gonna give it a try on my own branch and see if I can figure it out as well.

indragiek commented:
Don't worry about this, I managed to figure it out in #79.
